[IMPROVED] Audited README files for consistency, formatting, and style (#237)
1Password/scim-examples: docker/README.md

# Deploying the 1Password SCIM bridge using Docker
# Deploy 1Password SCIM Bridge with Docker

This example describes the methods of deploying the 1Password SCIM bridge using Docker. The Docker Compose and Docker Swarm managers are available and deployment using each manager is described below.
*Learn how to deploy 1Password SCIM Bridge using Docker Compose or Docker Swarm.*

## Preparing
**Table of contents:**

> Ensure you've read through the [PREPARATION.md](/PREPARATION.md) document before beginning deployment.

Reviewer: It might be nice to standardize this line across the deployment options. AWS says:

"Before beginning, familiarize yourself with [PREPARATION.md](/PREPARATION.md) and complete the necessary steps there."

- [Step 4: Test the SCIM bridge](#step-4-test-the-scim-bridge)
- [Step 5: Connect your identity provider](#step-5-connect-your-identity-provider)
- [Update your SCIM Bridge](#update-your-scim-bridge)
- [Appendix: Advanced `scim.env` options](#appendix-advanced-scimenv-options)
- [Appendix: Generate `scim.env` on Windows](#appendix-generate-scimenv-on-windows)

## Before you begin

> Before you begin, read the [PREPARATION.md](/PREPARATION.md) document.

Reviewer: Similarly, it might be nice to standardize this section. AWS says:

"Before you begin, familiarize yourself with [PREPARATION.md](/PREPARATION.md) and complete the necessary steps there."

## Step 1: Choose a deployment option

> Using Docker, you have two deployment options: Docker Compose and Docker Swarm.

Reviewer: It might be nice to link to the Docker pages for these here, so people can learn more about which option suits them better.

> **Docker Swarm** is the recommended option, but Docker Compose can also be used depending on your deployment needs. You can either set one up a Docker host on your own infrastructure or on a cloud provider of your choice.

Reviewer: ```suggestion
**Docker Swarm** is the recommended option, but Docker Compose can also be used depending on your deployment needs. You can set up a Docker host on your own infrastructure or on a cloud provider of your choice.
```

The `scimsession` file is passed into the docker container via an environment variable, which is less secure than Docker Swarm secrets, Kubernetes secrets, or AWS Secrets Manager, all of which are supported and recommended for production use.

## Step 2: Install Docker tools

1. On your local machine, install [Docker for Desktop](https://www.docker.com/products/docker-desktop) and start Docker.
2. Install the `docker-compose` and `docker-machine` command-line tools on your local machine. If you're using macOS and Homebrew, make sure you're using the _cask_ app-based version of Docker (`brew cask install docker`), not the default CLI version.

## Step 3: Deploy 1Password SCIM Bridge

To automatically deploy 1Password SCIM Bridge with [Docker Swarm](#docker-swarm) or [Docker Compose](#docker-compose), use our script, [./docker/deploy.sh](deploy.sh).

### Docker Swarm

For this method, you'll need to have joined a Docker Swarm with the target deployment node. Learn how to [create a swarm](https://docs.docker.com/engine/swarm/swarm-tutorial/create-swarm/).

> After you've created a swarm, log in with `docker swarm join`. Then use the provided the bash script [./docker/deploy.sh](deploy.sh) to deploy your SCIM bridge. The script will do the following:

Reviewer: ```suggestion
After you've created a swarm, log in with `docker swarm join`. Then use the provided bash script [./docker/deploy.sh](deploy.sh) to deploy your SCIM bridge. The script will do the following:
```

> 1. Prompt whether you're using Google Workspace as your identity provider so you can add your configuration files as Docker Secrets within your Swarm cluster.

Reviewer: ```suggestion
1. Ask whether you're using Google Workspace as your identity provider so you can add your configuration files as Docker Secrets within your Swarm cluster.
```

"Prompt whether" sounds a bit odd to me, for some reason. Assuming this is an interactive prompt, like it sounds like, I think "ask" would sound more natural for most of these.

> 2. Prompt whether you're deploying using Docker Swarm or Docker Compose.

Reviewer: ```suggestion
2. Ask whether you're deploying using Docker Swarm or Docker Compose.
```

> 3. Prompt for your SCIM bridge domain name so you can automatically get a TLS certificate from Let's Encrypt. This is the domain you selected in [PREPARATION.md](/PREPARATION.md).

Reviewer: ```suggestion
3. Ask for your SCIM bridge domain name so you can automatically get a TLS certificate from Let's Encrypt. This is the domain you selected in [PREPARATION.md](/PREPARATION.md).
```

> 4. Prompt for your `scimsession` file location to add your `scimsession` file as a Docker Secret within your Swarm cluster.

Reviewer: ```suggestion
4. Ask for your `scimsession` file location to add your `scimsession` file as a Docker Secret within your Swarm cluster.
```

5. Deploy a container using `1password/scim`, as well as a `redis` container. The `redis` container is necessary to store Let's Encrypt certificates, as well as act as a cache for your identity provider information.

The logs from the SCIM bridge and Redis containers will be streamed to your machine. If everything seems to have deployed successfully, press Ctrl+C to exit, and the containers will remain running on the remote machine.

At this point, you should set a DNS record routing the domain name to the IP address of the `op-scim` container.

### Docker Compose

To deploy with Docker Compose, you'll need Docker Desktop set up either locally or remotely. Learn how to [set up Docker Desktop](https://docs.docker.com/desktop/). Then follow these steps:

> 1. Make sure your environment is set up with `eval %{docker-machine env $machine_name}`, using the machine name you've chosen.

Reviewer: Possibly stupid question - is this a command that would be run on the command line? It might be nice to specify that, if so.

```suggestion
1. Make sure your environment is set up by running the command `eval %{docker-machine env $machine_name}` using the machine name you've chosen.
```

# (optional) view the container logs
docker-compose -f docker-compose.yml logs -f
```
Learn more about [connecting Google Workspace to 1Password SCIM Bridge](https://support.1password.com/scim-google-workspace/).

</details>

<hr>

## Step 4: Test the SCIM bridge

> To test if your SCIM bridge came online, open the public IP address of the Docker Host for your bridge in a web browser. You should be able to enter your bearer token to verify that your SCIM bridge is up and running.

Reviewer: ```suggestion
To test if your SCIM bridge is online, open the public IP address of the Docker Host for your bridge in a web browser. You should be able to enter your bearer token to verify that your SCIM bridge is up and running.
```
